### PR TITLE
Aave-v3-import with permit | Avalanche, Polygon, Arbitrum

### LINKED ISSUE #201: Aave v3 import connector with permit
### Current Implementation:
Aave v3 import connector: imports Aave v3 position from the user's EOA address to user's DSA wallet in a single transaction with some allowance transactions.

### Issue: 
User needs to give allowances of all supplied tokens i.e atokens to user's DSA wallet address before doing import transaction. So, if the user has 5 atokens, then 5 allowance transactions + 1 import transaction => total 6 transactions.

### Solution: 
Since Aave v3 atokens have permit functionality we can utilize and completely remove the allowance transactions and do permit message signing from the user EOA address and use those signed permits in import transaction

### Changes:
[Aave v3 import connector on polygon](https://github.com/Instadapp/dsa-connectors/blob/main/contracts/polygon/connectors/aave/v3-import/main.sol):

We need to create a similar Aave v3 import connector for polygon with permit functionality. So, in the import function, we will be passing extra param `signedPermits`. we need to add permit function before taking atokens(i.e [_TransferAtokens](https://github.com/Instadapp/dsa-connectors/blob/main/contracts/polygon/connectors/aave/v3-import/main.sol#L50)) from user's EOA to DSA wallet

### PR DESCRIPTION
**Pull request by iri031**
Implemented connector for aave-v3 import with Permit function to allow single transaction import, removing allowance transactions.

closes #201 

- [x] Added function in [helpers.sol](https://github.com/Richa-iitr/dsa-connectors/blob/aave-v3-import-permit/contracts/polygon/connectors/aave/v3-import-permit/helpers.sol#L233)
- [x] Testing 
- [x] Review and approve connector @thrilok209 @pradyuman-verma 
- [x] Deploy and comment addresses @richa-iitr
- [x] enable and add to SDK https://github.com/Instadapp/dsa-connect/pull/152
- [x] Add to docs https://github.com/Instadapp/docs/pull/161

**Test cases added for :** 
- Polygon
- Arbitrum
- Avalanche

### CODE REVIEW

File: test/polygon/aave/aaveV3-import-test.ts [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+
+  const daiAbi = abis.basic.erc20;
+
+  let aEth: Contract, aDai: Contract, Dai: any;
+  let dsaWallet0: any;
+  let masterSigner: Signer;
+  let instaConnectorsV2: Contract;
+  let connector: any;
+
+  const wallet = ethers.Wallet.fromMnemonic(mnemonic);
```
Comment (thrilok209) [inline test/polygon/aave/aaveV3-import-test.ts:219]: Why is wallet created like this? We can directly use it, right? Hardhat by default injects the same mnemonic in ethersjs

Comment (thrilok209) [reply test/polygon/aave/aaveV3-import-test.ts:219]: https://github.com/Instadapp/dsa-connectors/blob/main/test/mainnet/compound/compound.test.ts#L25-L26

Comment (pradyuman-verma) [reply test/polygon/aave/aaveV3-import-test.ts:219]: Yes, but to sign the message, we need a private key, which we can get by using the wallet. 

If we use signer to sign a message then the signed message will be having `0x19 Ethereum ...`

File: contracts/polygon/connectors/aave/v3-import-permit/helpers.sol [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+			}
+		}
+		for (uint256 i = 0; i < inputData.supplyTokens.length; i++) {
+			address _token = inputData.supplyTokens[i] == maticAddr
+				? wmaticAddr
+				: inputData.supplyTokens[i];
+			(address _aToken, , ) = aaveData.getReserveTokensAddresses(_token);
+			data._supplyTokens[i] = _token;
+			data.aTokens[i] = ATokenInterface(_aToken);
+			data.supplyAmts[i] = data.aTokens[i].balanceOf(userAccount);
```
Comment (pradyuman-verma) [inline contracts/polygon/connectors/aave/v3-import-permit/helpers.sol:189]: Remove this variable, and we will approve the maximum amount, i.e., unit(-1).

File: contracts/polygon/connectors/aave/v3-import-permit/helpers.sol [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+		ATokenInterface[] memory aTokenContracts,
+		address[] memory tokens,
+		uint256[] memory amts,
+		uint8[] memory v,
+		bytes32[] memory r,
+		bytes32[] memory s,
+		uint256[] memory expiry
+	) internal {
+		for(uint256 i = 0; i < tokens.length; i++) {
+			aTokenContracts[i].permit(userAccount, address(this), amts[i], expiry[i], v[i], r[i], s[i]);
```
Comment (pradyuman-verma) [inline contracts/polygon/connectors/aave/v3-import-permit/helpers.sol]:
```suggestion
			aTokenContracts[i].permit(userAccount, address(this), uint(-1), expiry[i], v[i], r[i], s[i]);
```

File: contracts/polygon/connectors/aave/v3-import-permit/helpers.sol (hunk 2 above)
Comment (iri031) [reply contracts/polygon/connectors/aave/v3-import-permit/helpers.sol:189]: I need to remove the supplyAmts?

File: test/polygon/aave/aaveV3-import-test.ts [hunk 4]
```diff
@@ -25,13 +25,237 @@ const ABI = [
...
+      {
+        "indexed": false,
+        "name": "value",
+        "type": "uint256"
+      }
+    ],
+    "name": "Transfer",
+    "type": "event"
+  }
+];
```
Comment (pradyuman-verma) [inline test/polygon/aave/aaveV3-import-test.ts:206]: Why create this?

Comment (iri031) [reply test/polygon/aave/aaveV3-import-test.ts:206]: The IERC20__factory which was generated in my case, didn't contain all the methods, so i used this for checking.

**Comment by iri031** [comment]
Screenshot from 2022-04-04 01-40-37

File: contracts/polygon/connectors/aave/v3-import-permit/helpers.sol [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+		bool convertStable;
+		uint256[] flashLoanFees;
+	}
+
+	struct signedPermits {
+		uint8[] v;
+		bytes32[] r;
+		bytes32[] s;
+		uint256[] expiry;
+	}
```
Comment (thrilok209) [inline contracts/polygon/connectors/aave/v3-import-permit/helpers.sol]:
```suggestion
	struct SignedPermits {
		uint8[] v;
		bytes32[] r;
		bytes32[] s;
		uint256[] expiry;
	}
```

File: contracts/polygon/connectors/aave/v3-import-permit/main.sol [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+import { TokenInterface, AccountInterface } from "../../../common/interfaces.sol";
+import { AaveInterface, ATokenInterface } from "./interface.sol";
+import "./helpers.sol";
+import "./events.sol";
+
+contract AaveV3ImportPermitResolver is AaveHelpers {
+	function _importAave(
+		address userAccount,
+		ImportInputData memory inputData,
+		signedPermits memory permitData
```
Comment (thrilok209) [inline contracts/polygon/connectors/aave/v3-import-permit/main.sol]:
```suggestion
		SignedPermits memory permitData
```

File: contracts/polygon/connectors/aave/v3-import-permit/main.sol [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+	 * @dev Import aave V3 position .
+	 * @notice Import EOA's aave V3 position to DSA's aave v3 position
+	 * @param userAccount The address of the EOA from which aave position will be imported
+	 * @param inputData The struct containing all the neccessary input data
+	 * @param permitData The struct containing signed permit data like v,r,s,expiry
+	 */
+	function importAave(
+		address userAccount,
+		ImportInputData memory inputData,
+		signedPermits memory permitData
```
Comment (thrilok209) [inline contracts/polygon/connectors/aave/v3-import-permit/main.sol]:
```suggestion
		SignedPermits memory permitData
```

File: test/polygon/aave/aaveV3-import-test.ts [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+
+const token = new ethers.Contract(DAI, erc20Abi);
+const aDai = new ethers.Contract(aDaiAddress, ABI);
+const usdcToken = new ethers.Contract(USDC, erc20Abi);
+
+describe("Import Aave", function () {
+  const connectorName = "AAVE-V3-IMPORT-PERMIT-X";
+  let signer: any, wallet0: any;
+
+  const aaveAbi = [
```
Comment (thrilok209) [inline test/polygon/aave/aaveV3-import-test.ts]: @Richa-iitr I think we can declare this outside of `describe`?

File: test/polygon/aave/aaveV3-import-test.ts [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+    stateMutability: "nonpayable",
+    type: "function"
+  }
+];
+
+const token = new ethers.Contract(DAI, erc20Abi);
+const aDai = new ethers.Contract(aDaiAddress, ABI);
+const usdcToken = new ethers.Contract(USDC, erc20Abi);
+
+describe("Import Aave", function () {
```
Comment (thrilok209) [inline test/polygon/aave/aaveV3-import-test.ts]:
```suggestion
describe("Import Aave V3 with Permit", function () {
```

File: test/polygon/aave/aaveV3-import-test.ts [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+    // approve DAI to aavePool
+    await token.connect(wallet0).approve(aaveAddress, parseEther("8"));
+
+    //deposit DAI in aave
+    await aave.connect(wallet0).supply(DAI, parseEther("8"), wallet.address, 3228);
+    console.log("Supplied DAI on aave");
+
+    //borrow USDC from aave
+    await aave.connect(wallet0).borrow(USDC, parseUnits("5", 6), 2, 3228, wallet.address);
+    console.log("Borrowed USDC from aave");
```
Comment (thrilok209) [inline test/polygon/aave/aaveV3-import-test.ts]: @Richa-iitr Move this to another `it` hook? Like "Should create Aave v3 position of DAI(collateral) and USDC(debt)"

Comment (thrilok209) [reply test/polygon/aave/aaveV3-import-test.ts]: Also, add `expect` to check DAI collateral balance and USDC debt balance.

File: test/polygon/aave/aaveV3-import-test.ts [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+  describe("check user AAVE position", async () => {
+    it("Should check position of user", async () => {
+      expect(await aDai.connect(wallet0).balanceOf(wallet.address)).to.be.gte(
+        new BigNumber(8).multipliedBy(1e18).toString()
+      );
+
+      expect(await usdcToken.connect(wallet0).balanceOf(wallet.address)).to.be.gte(
+        new BigNumber(5).multipliedBy(1e6).toString()
+      );
+    });
```
Comment (thrilok209) [inline test/polygon/aave/aaveV3-import-test.ts]: @Richa-iitr Ahh! okay. Maybe check this after creating Aave v3 position

**Comment by iri031** [comment]
Screenshot from 2022-04-04 21-26-59

File: test/polygon/aave/aaveV3-import-test.ts (hunk 8 above)
Comment (iri031) [reply test/polygon/aave/aaveV3-import-test.ts]: Done the changes

File: test/polygon/aave/aaveV3-import-test.ts (hunk 10 above)
Comment (iri031) [reply test/polygon/aave/aaveV3-import-test.ts]: Done!

File: test/polygon/aave/aaveV3-import-test.ts (hunk 11 above)
Comment (iri031) [reply test/polygon/aave/aaveV3-import-test.ts]: Made the changes.

File: contracts/polygon/connectors/aave/v3-import-permit/events.sol [hunk 12]
```diff
@@ -0,0 +1,15 @@
+//SPDX-License-Identifier: MIT
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+contract Events {
+	event LogAaveV3Import(
```
Comment (pradyuman-verma) [inline contracts/polygon/connectors/aave/v3-import-permit/events.sol]:
```suggestion
	event LogAaveV3ImportWithPermit(
```

File: contracts/polygon/connectors/aave/v3-import-permit/main.sol [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+			);
+			_BorrowVariable(
+				data._borrowTokens.length,
+				aave,
+				data._borrowTokens,
+				data.variableBorrowAmtsWithFee
+			);
+		}
+
+		_eventName = "LogAaveV3Import(address,bool,address[],address[],uint256[],uint256[],uint256[],uint256[])";
```
Comment (pradyuman-verma) [inline contracts/polygon/connectors/aave/v3-import-permit/main.sol]:
```suggestion
		_eventName = "LogAaveV3ImportWithPermit(address,bool,address[],address[],uint256[],uint256[],uint256[],uint256[])";
```

File: contracts/polygon/connectors/aave/v3-import-permit/events.sol [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+		address indexed user,
+		address[] ctokens,
```
Comment (pradyuman-verma) [inline contracts/polygon/connectors/aave/v3-import-permit/events.sol]:
```suggestion
		address[] atokens,
```

**Review by pradyuman-verma — DISMISSED** [review]

**Comment by iri031** [comment]
Arbitrum chain
Screenshot from 2022-04-06 01-38-30

**Review by pradyuman-verma — DISMISSED** [review]

**Comment by iri031** [comment]
Avalanche chain
Screenshot from 2022-04-06 03-53-35

**Review by pradyuman-verma — DISMISSED** [review]

**Comment by iri031** [comment]
## Deployed connector for Polygon 

**Name** `AAVE-V3-IMPORT-PERMIT-A`
**Addr** [0xe6aa2d277aAfBb9e19354F6f893737C3608ff995](https://polygonscan.com/address/0xe6aa2d277aAfBb9e19354F6f893737C3608ff995#code)
**tx** [here](https://polygonscan.com/tx/0x2e529e6fef4f0f3147b79124e74bfc69470e031aebb552d2276b2a5ef47a4a84)

## Deployed connector for Avalanche
**Name** `AAVE-V3-IMPORT-PERMIT-A`
**Addr** [0x2249Afe319baA0137CB434116Ac4d4Fd5c3eb719](https://snowtrace.io/address/0x2249Afe319baA0137CB434116Ac4d4Fd5c3eb719#code)
**tx** [here](https://snowtrace.io/tx/0xbd314ae38d3cf802b38d99a43c7d135f7c61762eeb804061fe917c3ffde05a13)

## Deployed connector for Arbitrum
**Name** `AAVE-V3-IMPORT-PERMIT-A`
**Addr** [0xe6aa2d277aAfBb9e19354F6f893737C3608ff995](https://arbiscan.io/address/0xe6aa2d277aAfBb9e19354F6f893737C3608ff995#code)
**tx** [here](https://arbiscan.io/tx/0xe5115f8e687aa3bb1f164a6aee0ad602cc6e52af8b7434ab5f815ebea9864c71)

**Review by pradyuman-verma — APPROVED** [review]

**Review by thrilok209 — APPROVED** [review]

**Pull request merged**